File: Parallel and Distributed Programming/Lab0-Java/src/main/java/Adunare/Calculator.java
package Adunare;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Calculator
{
    private List<Integer> numarMare1;
    private List<Integer> numarMare2;

    public Calculator(List<Integer> numarMare1, List<Integer> numarMare2) {
        this.numarMare1 = numarMare1;
        this.numarMare2 = numarMare2;

        this.checkVectors();

        Collections.reverse(numarMare1);
        Collections.reverse(numarMare2);

    }

    private void checkVectors(){
        while (numarMare1.size() > numarMare2.size())
            numarMare2.add(0);

        while (numarMare2.size() > numarMare1.size())
            numarMare1.add(0);
    }

    private String operatorX(String x, String y){
        if(x.equals("N") && y.equals("N"))
            return "N";
        else if(x.equals("N") && y.equals("M"))
            return "N";
        else if(x.equals("N") && y.equals("C"))
            return "C";
        else if(x.equals("M") && y.equals("N"))
            return "N";
        else if(x.equals("M") && y.equals("M"))
            return "M";
        else if(x.equals("M") && y.equals("C"))
            return "C";
        else if(x.equals("C") && y.equals("N"))
            return "N";
        else if(x.equals("C") && y.equals("M"))
            return "C";
        else
            return "C";
    }

    private List<String> pas1(){
        List<String> c = new ArrayList<>();
        for (int i = 0; i < this.numarMare1.size(); i++){
            if(numarMare1.get(i) + numarMare2.get(i) >= 10)
                c.add("C");

            if(numarMare1.get(i) + numarMare2.get(i) == 9)
                c.add("M");

            if(numarMare1.get(i) + numarMare2.get(i) <= 8)
                c.add("N");
        }
        return c;
    }

    private List<String> pas2(){
        List<String> c = this.pas1();
        List<String> e = new ArrayList<>();
        for(int i = 0; i < c.size(); i++){
            if(i == 0){
                e.add(c.get(i));
            } else
            if(c.get(i).equals("M"))
            {
                int j = i-1;
                while(c.get(j).equals("M") && j > 0){
                    j--;
                }
                if(c.get(j).equals("C"))
                    e.add("C");
                else e.add(this.operatorX(c.get(i - 1), c.get(i)));
            }
            else e.add(this.operatorX(c.get(i - 1), c.get(i)));
        }
        return e;
    }

    private List<Integer> pas3(){
        List<Integer> m = new ArrayList<>();
        List<String> e = pas2();
        for(int i = 0; i < e.size(); i++){
            if(i == 0){
                m.add(0);
            }else
                if(e.get(i - 1).equals("C"))
                    m.add(1);
                else
                    m.add(0);
        }
        return m;
    }

    private List<Integer> pas4(){
        List<Integer> r = new ArrayList<>();
        List<Integer> m =pas3();
        for(int i = 0; i < m.size(); i++){
            r.add((numarMare1.get(i) + numarMare2.get(i) + m.get(i))%10);
            if(i == m.size() - 1){
                int b = (numarMare1.get(i) + numarMare2.get(i) + m.get(i))%100;
                if(b > 9){
                    r.add(b / 10);
                }
            }
        }

        return r;
    }

    public List<Integer> aduna(){
        List<Integer> r = pas4();
        Collections.reverse(r);
        return r;
    }
}

File: Parallel and Distributed Programming/Lab0-C++/Lab0-C++/Calculator.cpp
#include "Calculator.h"

void Calculator::checkVectors() {
	while (numarMare1.size() > numarMare2.size())
		numarMare2.push_back(0);
	while (numarMare1.size() < numarMare2.size())
		numarMare1.push_back(0);
}

std::string Calculator::operatorX(std::string x, std::string y) {
	if (x == "N" && y == "N")
		return "N";
	else if (x == "N" && y == "M")
		return "N";
	else if (x == "N" && y =="C")
		return "C";
	else if (x == "M" && y == "N")
		return "N";
	else if (x == "M" && y == "M")
		return "M";
	else if (x == "M" && y == "C")
		return "C";
	else if (x == "C" && y == "N")
		return "N";
	else if (x == "C" && y == "M")
		return "C";
	else
		return "C";
}

std::vector<std::string> Calculator::pas1() {
	std::vector<std::string> c;
	for (int i = 0; i < numarMare1.size(); i++) {
		if (numarMare1.at(i) + numarMare2.at(i) >= 10)
			c.push_back("C");

		if (numarMare1.at(i) + numarMare2.at(i) == 9)
			c.push_back("M");

		if (numarMare1.at(i) + numarMare2.at(i) <= 8)
			c.push_back("N");
	}
	return c;
}

std::vector<std::string> Calculator::pas2() {
	std::vector<std::string> c = pas1();
	std::vector<std::string> e;
	for (int i = 0; i < c.size(); i++) {
		if (i == 0) {
			e.push_back(c.at(i));
		}
		else
			if (c.at(i) == "M")
			{
				int j = i - 1;
				while (c.at(j) == "M" && j > 0) {
					j--;
				}
				if (c.at(j) == "C")
					e.push_back("C");
				else e.push_back(operatorX(c.at(i - 1), c.at(i)));
			}
			else e.push_back(operatorX(c.at(i - 1), c.at(i)));
	}
	return e;
}

std::vector<int> Calculator::pas3() {
	std::vector<int> m;
	std::vector<std::string> e = pas2();
	for (int i = 0; i < e.size(); i++) {
		if (i == 0) {
			m.push_back(0);
		}
		else
			if (e.at(i - 1) == "C")
				m.push_back(1);
			else
				m.push_back(0);
	}
	return m;
}

std::vector<int> Calculator::pas4() {
	std::vector<int> r;
	std::vector<int> m = pas3();
	for (int i = 0; i < m.size(); i++) {
		r.push_back((numarMare1.at(i) + numarMare2.at(i) + m.at(i)) % 10);
		if (i == m.size() - 1) {
			int b = (numarMare1.at(i) + numarMare2.at(i) + m.at(i)) % 100;
			if (b > 9) {
				r.push_back(b / 10);
			}
		}
	}

	return r;
}

std::vector<int> Calculator::aduna() {
	std::vector<int> r = pas4();
	std::reverse(std::begin(r), std::end(r));
	return r;
}